pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -1238,10 +1249,39 @@ class RuntimeASTTransformer {
             // These variable handles are declared in functions, but will still be bound to the script context
             // Doing this here ensures the variables make the trip across to run time - if they don't make it, neither did this pipeline
             pipelineBlock.addStatement(stmt(callX(closureX(currentBlock), 'call')))
-            pipelineElementHandles.clear()
         }
 
-        /**
+        @NonNull
+        private List<Statement> prepareClosureScopedHandles(@NonNull BlockStatement pipelineBlock) {
+            ArrayList<Statement> result = new ArrayList<Statement>()
+            if (SCRIPT_SPLITTING_TRANSFORMATION) {
+                ArrayList<DeclarationExpression> declarations = new ArrayList<DeclarationExpression>()
+                moduleNode.statementBlock.statements.each { item ->
+                    if (((ExpressionStatement) item).expression instanceof DeclarationExpression) {
+                        declarations.add((DeclarationExpression) ((ExpressionStatement) item).expression)
+                    }
+                }
+
+                if (!declarations.isEmpty()) {
+                    if (SCRIPT_SPLITTING_ALLOW_LOCAL_VARIABLES) {
+                        result.addAll(pipelineElementHandles)
+                        pipelineElementHandles.clear()
+                    } else {
+                        throw new IllegalStateException("[JENKINS-34987] SCRIPT_SPLITTING_TRANSFORMATION is an experimental feature of Declarative Pipeline and is incompatible with local variable declarations inside a Jenkinsfile. " +
+                                "Local variable declarations found: " + declarations.join(", ") + ". "
```

Comment:
I think the compilation error is because you are missing a `+` here:

```suggestion
                                "Local variable declarations found: " + declarations.join(", ") + ". " +
```

Reply:
Dang it, I knew I was going to screw that up.